File: src/2025-03-04-12-20-21--deepseek-vr-0304/cpp/exercises/practice/phone-number/phone_number.cpp
#include "phone_number.h"
#include <cctype>

namespace phone_number {

PhoneNumber::PhoneNumber(const std::string& s) {
    std::string cleaned;
    for (char c : s) {
        if (std::isdigit(static_cast<unsigned char>(c))) {
            cleaned += c;
        }
    }
    
    // Check country code and length
    if (cleaned.size() == 11 && cleaned[0] == '1') {
        cleaned = cleaned.substr(1);
    } else if (cleaned.size() != 10) {
        throw std::invalid_argument(cleaned.empty() ? "Empty input" : "Invalid number");
    }
    
    // Validate area code (NXX) and exchange code (NXX)
    if (cleaned[0] < '2' || cleaned[0] > '9') {
        throw std::invalid_argument("Area code cannot start with " + std::string(1, cleaned[0]));
    }
    if (cleaned[3] < '2' || cleaned[3] > '9') {
        throw std::invalid_argument("Exchange code cannot start with " + std::string(1, cleaned[3]));
    }
    
    num_ = cleaned;
}

std::string PhoneNumber::number() const {
    return num_;
}

std::string PhoneNumber::area_code() const {
    return num_.substr(0, 3);
}

}  // namespace phone_number

File: src/2025-03-04-12-20-21--deepseek-vr-0304/javascript/exercises/practice/phone-number/phone-number.js
//
// This is only a SKELETON file for the 'Phone Number' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export const clean = (number) => {
  // Check for letters
  if (/[a-zA-Z]/.test(number)) {
    throw new Error('Letters not permitted');
  }

  // Check for invalid punctuation
  if (/[^\d\s().+-]/.test(number)) {
    throw new Error('Punctuations not permitted');
  }

  // Clean the number by removing all non-digits
  let cleaned = number.replace(/[^\d]/g, '');

  // Check length and country code
  if (cleaned.length === 11) {
    if (cleaned.startsWith('1')) {
      cleaned = cleaned.substring(1);
    } else {
      throw new Error('11 digits must start with 1');
    }
  }

  // Check number of digits
  if (cleaned.length !== 10) {
    throw new Error('Incorrect number of digits');
  }

  // Check area code
  if (cleaned[0] === '0') {
    throw new Error('Area code cannot start with zero');
  }
  if (cleaned[0] === '1') {
    throw new Error('Area code cannot start with one');
  }

  // Check exchange code
  if (cleaned[3] === '0') {
    throw new Error('Exchange code cannot start with zero');
  }
  if (cleaned[3] === '1') {
    throw new Error('Exchange code cannot start with one');
  }

  return cleaned;
};
